src/lab16.py

```diff
@@ -278,69 +308,49 @@ def __init__(self, node, parent, previous, frame):
         self.y = None
         self.width = None
         self.height = None
+        self.zoom = None
 
         self.dirty_children = True
-        self.dirty_zoom = True
-        self.zoom = None
+        self.zoom_field = DependentField(self, "zoom")
         self.dirty_style = True
-        self.dirty_width = True
-        self.dirty_height = True
-        self.dirty_x = True
-        self.dirty_y = True
+        self.width_field = DependentField(self, "width")
+        self.x_field = DependentField(self, "x")
+        self.y_field = DependentField(self, "y")
+        self.height_field = DependentField(self, "height")
         self.dirty_descendants = True
 
     def layout(self):
-        if self.dirty_zoom:
-            self.zoom = self.parent.zoom
-            for child in self.children:
-                mark_dirty(child)
-                child.dirty_zoom = True
-            mark_dirty(self)
-            self.dirty_width = True
-            self.dirty_zoom = False
+        parent_zoom = self.zoom_field.read(self.parent.zoom_field)
+        self.zoom = self.zoom_field.set(parent_zoom)
 
         if self.dirty_style:
             self.dirty_children = True
-            self.dirty_width = True
+            self.width_field.dirty = True
             mark_dirty(self)
             self.dirty_style = False
 
-        if self.dirty_width:
-            assert not self.parent.dirty_width
-            if "width" in self.node.style:
-                self.width = device_px(float(self.node.style["width"][:-2]), zoom)
-            else:
-                self.width = self.parent.width
+        assert not self.dirty_style
+        old_w = self.width
+        if "width" in self.node.style:
+            zoom = self.width_field.read(self.zoom_field)
+            self.width = self.width_field.set(float(self.node.style["width"][:-2], zoom))
+        else:
+            parent_width = self.width_field.read(self.parent.width_field)
+            self.width = self.width_field.set(parent_width)
+        if self.width != old_w:
             self.dirty_children = True
-            for child in self.children:
```

Comment:
Am I right that we get this for free because the first time through we recurse into children, and they mark themselves as depending on their parent?

Reply:
Yes, exactly. Every field remembers who read it, and dirties those on change.